src/main/java/me/kavin/piped/server/ServerLauncher.java

```diff
@@ -293,6 +304,88 @@ AsyncServlet mainServlet(Executor executor) {
+                                        URI.create(request.getFullUrl())
+                                );
+
+                                if (response instanceof AuthenticationErrorResponse) {
+                                    // The OpenID provider returned an error
+                                    System.err.println(response.toErrorResponse().getErrorObject());
+                                    return HttpResponse.ofCode(500).withHtml("OpenID provider returned an error:\n\n" + response.toErrorResponse().getErrorObject().toString());
+                                }
+                                AuthenticationSuccessResponse sr = response.toSuccessResponse();
+
+                                AuthorizationCode code = sr.getAuthorizationCode();
+                                AuthorizationGrant codeGrant = new AuthorizationCodeGrant(
+                                        code, callback
+                                );
+
+                                TokenRequest tr = new TokenRequest(provider.tokenUri, clientAuth, codeGrant);
+                                TokenResponse tokenResponse = OIDCTokenResponseParser.parse(tr.toHTTPRequest().send());
+
+                                if (! tokenResponse.indicatesSuccess()) {
+                                    TokenErrorResponse errorResponse = tokenResponse.toErrorResponse();
+                                    return HttpResponse.ofCode(500).withHtml("Failure while trying to request token:\n\n" + errorResponse.getErrorObject().getDescription());
+                                }
+
+                                OIDCTokenResponse successResponse = (OIDCTokenResponse)tokenResponse.toSuccessResponse();
+
+
+                                UserInfoRequest ur = new UserInfoRequest(provider.userinfoUri, successResponse.getOIDCTokens().getBearerAccessToken());
+                                UserInfoResponse userInfoResponse = UserInfoResponse.parse(ur.toHTTPRequest().send());
+
+                                if (! userInfoResponse.indicatesSuccess()) {
+                                    System.out.println(userInfoResponse.toErrorResponse().getErrorObject().getCode());
+                                    System.out.println(userInfoResponse.toErrorResponse().getErrorObject().getDescription());
+                                    return HttpResponse.ofCode(500).withHtml("Failed to query userInfo:\n\n" + userInfoResponse.toErrorResponse().getErrorObject().getDescription());
+                                }
+
+                                UserInfo userInfo = userInfoResponse.toSuccessResponse().getUserInfo();
+
+                                String sessionId = UserHandlers.oidcCallbackResponse(provider.name, userInfo.getSubject().toString());
+
+                                return HttpResponse.redirect302(Constants.FRONTEND_URL + "/login?session=" + sessionId);

```

Comment:
Rather than doing this, does OIDC support something like a `return_url` parameter or something similar? This way it can be used in any frontend, and in external applications like Libretube too.

Reply:
Im really not sure if this is something thats possible. The proper way to to this would be to use the PKCE flow, but we can't do that because it would bypass the backend thus skipping creation of the session token etc. We can add a redirect parameter to the login path, but with my provider (zitadel) this would require whitelisting the return url for every client manually.
Another alternative is to use some kind of otp to allow clients to receive a session token outside of the usual login process. Jellyfin does it this way (although they don't support oidc out of the box). The flow would look something like this:
1. User logs into website
2. User requests new otp from settings
3. User enters email + otp in external client
4. Backend checks if otp is valid and sends back session token

Reply:
> this would require whitelisting the return url for every client manually

What if we return to the API URL, then redirect from there based on a parameter passed?

Reply:
Since the redirect to the api url  including the query parameters is decided by the oidc provider we can't control the parameters unless we include them in the original requested callback uri, thus requiring the whitelist.

Reply:
There is no workaround for this since the code flow is specifically designed to not allow arbitrary redirect urls. With the PKCE flow this would work, but we would have to completely change the backend authentication and I probably would need to implement it from scratch since the oidc library doesn't support it.

Reply:
We could add another additional prompt with a warning for this, so they need to check the target redirected URL before proceeding.